4 run the tests in gh actions

rosepearson opened the pull request:
PR for using git actions to install Python dependencies, run tests and lint with a variety of Python versions.

Requirements prior to landing:
1. ~~At least one test passing~~
2. ~~Run all tests~~
3. ~~Create a strict and less strict test for windows and Linux~~
4. ~~Tidy up workflow file~~
5. ~~Ensure caching is working correctly~~

kinow commented on setup.py:
```diff
@@ -40,13 +40,17 @@
+                      'scipy==1.6.3',
+                      'requests==2.25.1',
+                      'boto3==1.17.105',
+                      'matplotlib==3.4.2'],
```
These versions can be relaxed a bit I think, e.g. was `'numpy>=1.21'`, now it could be `'numpy==1.21.*'` so that users automatically get the latest bug fixes...

rosepearson commented:
Yes – definitely. This commit is only for exploring why I was getting errors when using pip git install in a conda environment – as this is the first package I’ve used this in and I am/was surprised about the install errors I was getting. I just thought I would try nail things exactly to a conda environment where the packages were installed successfully together. Anyway – I think there will be a few more spurious commits to come – I’ll move the PR from draft once I’ve moved beyond the exploration stage and addressed the issue properly 😊

From: Bruno P. Kinoshita ***@***.***>
Sent: Monday, 19 July 2021 11:30 am
To: rosepearson/GeoFabrics ***@***.***>
Cc: Rose Pearson ***@***.***>; Author ***@***.***>
Subject: Re: [rosepearson/GeoFabrics] 4 run the tests in gh actions (#14)


@kinow commented on this pull request.

________________________________

In setup.py<https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frosepearson%2FGeoFabrics%2Fpull%2F14%23discussion_r671913887&data=04%7C01%7Crose.pearson%40niwa.co.nz%7C976b464981624b5e7a5708d94a43e63b%7C41caed736a0c468aba499ff6aafd1c77%7C0%7C0%7C637622477788417868%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=dYr8Bq%2F%2FSmiPdVbfb2f%2BVTc9uZ8sjZNPs3Z2mAPc5fw%3D&reserved=0>:

> -                      'pdal>=2.2',

-                      'geopandas>=0.9',

-                      'shapely>=1.7',

-                      'scipy>=1.6',

-                      'pathlib>=1.0'],

+    install_requires=['pyproj==3.0.1',

+                      'numpy==1.21.0',

+                      'rioxarray==0.4.2',

+                      'pdal==2.2.0',

+                      'python-pdal==2.4.0',

+                      'geopandas==0.9.0',

+                      'shapely==1.7.1',

+                      'scipy==1.6.3',

+                      'requests==2.25.1',

+                      'boto3==1.17.105',

+                      'matplotlib==3.4.2'],

These versions can be relaxed a bit I think, e.g. was 'numpy>=1.21', now it could be 'numpy==1.21.*' so that users automatically get the latest bug fixes...

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frosepearson%2FGeoFabrics%2Fpull%2F14%23pullrequestreview-709071901&data=04%7C01%7Crose.pearson%40niwa.co.nz%7C976b464981624b5e7a5708d94a43e63b%7C41caed736a0c468aba499ff6aafd1c77%7C0%7C0%7C637622477788427825%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=%2BF6pMmc4lUFb%2BQW0Vy%2BIJ%2B9GU17BtcubDsBZ%2FyFvRkE%3D&reserved=0>, or unsubscribe<https://aus01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFOS4FGHYS4FUEVL3YHGOITTYNPVVANCNFSM5ASS7E2A&data=04%7C01%7Crose.pearson%40niwa.co.nz%7C976b464981624b5e7a5708d94a43e63b%7C41caed736a0c468aba499ff6aafd1c77%7C0%7C0%7C637622477788427825%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=91WW%2BrYG8YI7pD7x3zmFPK2g1Os2AQNBDiNqull6m3k%3D&reserved=0>.

[https://www.niwa.co.nz/static/niwa-2018-horizontal-180.png] <https://www.niwa.co.nz>
Dr Rose Pearson
Data Scientist

+64-3-343-7862

National Institute of Water & Atmospheric Research Ltd (NIWA)
10 Kyle Street, Riccarton Christchurch New Zealand
Connect with NIWA: niwa.co.nz<https://www.niwa.co.nz> Facebook<https://www.facebook.com/nzniwa> LinkedIn<https://www.linkedin.com/company/niwa> Twitter<https://twitter.com/niwa_nz> Instagram<https://www.instagram.com/niwa_science>
To ensure compliance with legal requirements and to maintain cyber security standards, NIWA's IT systems are subject to ongoing monitoring, activity logging and auditing. This monitoring and auditing service may be provided by third parties. Such third parties can access information transmitted to, processed by and stored on NIWA's IT systems

kinow commented:
Still lurking here! Just wanted to point out that I think you are having the exact same error I am. Your version of Ubuntu on GH actions is the same as mine (go to the latest job, then the setup step, expand the operating system section: `20.04.2`).

The "latest" Ubuntu in GH actions means the latest LTS: https://github.com/actions/virtual-environments

For Docker images, it's also good practice to point latest to a stable release, and not the latest-latest, which can be misleading. You should be able to use the latest-latest via Docker I think.

kinow commented:
And here's what I used to check the `pdal` library available for each Ubuntu: https://launchpad.net/ubuntu/+source/pdal, mine focal has only `2.0.1+ds-1build3`.

I think the required C++ lib version for Python `pdal-2.2.0`, for instance, is defined [here](https://github.com/PDAL/PDAL/blob/a3dceb4eb4fd2429fdcebf3de5fda83875f48bed/CMakeLists.txt#L8) (I think?).

rosepearson reviewed (commented):
> Still lurking here! Just wanted to point out that I think you are having the exact same error I am. Your version of Ubuntu on GH actions is the same as mine (go to the latest job, then the setup step, expand the operating system section: `20.04.2`).
> 
> The "latest" Ubuntu in GH actions means the latest LTS: https://github.com/actions/virtual-environments
> 
> For Docker images, it's also good practice to point latest to a stable release, and not the latest-latest, which can be misleading. You should be able to use the latest-latest via Docker I think.

Thanks for the comment! I sort of hoped you might take interest in my failing. Cool - I'd wondered if that was the case RE my error matching yours. I'm familiar with the principle of Docker, but less familiar with setting them up in practice. 

Would you recommend I try a different virtual-environment? I could try windows for instance - that is what I have on my local machine. When it comes to pinning to stable release, I guess you just google the package and take the latest stable that is supported by the OS? Thanks!

kinow commented:
>Would you recommend I try a different virtual-environment?

Let me see if I can find an example with Docker + GH Actions. We have one in Cylc Flow that runs the tests against 4 different versions of `bash`, but I will look for one for Python.

>I could try windows for instance - that is what I have on my local machine.

I think it's important to test -at least- in the environment where most of your users will be using. If most of the users of GeoFabrics will be using Windows, then I think it'd make sense to focus on using GH Actions + Windows (I never tried that, so I'd find it interest!). But if you are not sure, or if most users will be on Linux, then maybe we should try to finish the current GH Actions + ubuntu pipeline.

>When it comes to pinning to stable release, I guess you just google the package and take the latest stable that is supported by the OS? Thanks!

It depends I think. But normally I guess you want the most up to date environment possible. Given that that version is compatible with the other dependencies in the operating system. Sometimes you have to wait for Ubuntu, Red Hat, or even Windows, to release patches with updates for related packages before you can update lower level tools and libraries like `gcc`, `gdal`, `pdal`, etc.

kinow commented on .github/workflows/python-test-package.yml:
```diff
@@ -0,0 +1,62 @@
+on:
+  push:
+    branches:
+      - conda-1
```
Feel free to revert this to what you had before @rosepearson , so that we can trigger it for your PR now.

kinow commented on .github/workflows/python-test-package.yml:
```diff
@@ -0,0 +1,58 @@
+
+name: Python package using marketplace setup-miniconda
+
+on: [push, workflow_dispatch]
```
I think we need `pull_request` too? So that the pull requests are tested as well.

LukeParky replied:
You could add a status check to see branch is up to date before allowing merge. This would mean each PR is tested through the push workflow

kinow commented on .github/workflows/python-test-package.yml:
```diff
@@ -0,0 +1,58 @@
+        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
+    - name: Test with pytest
+      run: |
+        pytest tests/test_lidar_fetch/test_lidar_fetch.py
```
Missing newline, not too important.

kinow commented on setup.py:
```diff
@@ -42,11 +42,12 @@
+                      'boto3>=1.17'],
     extras_require={
         'dev': ['check-manifest'],
     },
```
Now that GH actions is using Conda, these changes and the ones in `requirements.txt` can be kept or reverted. If they are fixing/improving, then no harm in leaving them here. This could be useful if GeoFabrics is released to PYPI in the future.

kinow reviewed (approved):
Tests passing, CI running in a reasonable time, it's also printing info about the system. I think commits can be squashed if you'd like @rosepearson , but other than that this PR looks good to me! :tada: well done fixing the CI pipeline :1st_place_medal:

LukeParky commented on .github/workflows/python-test-package.yml:
```diff
@@ -0,0 +1,58 @@
+        use-only-tar-bz2: true # IMPORTANT: This needs to be set for caching to work properly!
+        python-version: ${{ matrix.python-version }}
+        mamba-version: "*"
+        # condarc-file: etc/example-condarc.yml
```
remove unused code
```suggestion
```

rosepearson replied:
good suggestion!

LukeParky commented on setup.py:
```diff
@@ -42,11 +42,12 @@
     install_requires=['numpy>=1.21',
                       'rioxarray>=0.4',
-                      'pdal>=2.2',
+                      'pdal>=2.2'
```
```suggestion
                      'pdal>=2.2',
```

LukeParky reviewed (changes requested):
Great that you got this working, I might pick your brain about it next time you're in at the GRI.
Depending on your pull request strategy you could enforce branches being up to date for a pull request.
Also, these commits would be the perfect candidate for squashing.

Good to go after some tiny suggested changes

rosepearson commented:
Awesome - Thanks @LukeParky and @kinow. I've actually still got to get the last two tests off the ground - so issues manifesting as file path errors but possibly related to something in the processor/rioxarray side of the dependency tree. I'll flick a message here once I've dealt with this and addressed your suggestions.

kinow commented:
@rosepearson 

>I've actually still got to get the last two tests off the ground - so issues manifesting as file path errors but possibly related to something in the processor/rioxarray side of the dependency tree.

I had errors yesterday when running the tests in the Docker image, and fixed in this commit: https://github.com/kinow/GeoFabrics/commit/e6425586b3ae08e1fa022a6efe6ef50123327843

There's a JSON file with backward slashes for path separators, which works on Windows and I thought would work with Python's `pathlib`, but for some strange reason created a file instead of a directory (a single file with the whole name `part\part\part\etc`).

Just in case you are seeing the same error...

rosepearson commented:
The test_processor_remote_tiles test is failing as the pdal writers.gdal is creating an offset dem - it overrides the exact grid location you specify with one that is nearby. This means the dem is not the same. I will see what happens if I use the same version of gdal.

rosepearson commented:
> The test_processor_remote_tiles test is failing as the pdal writers.gdal is creating an offset dem - it overrides the exact grid location you specify with one that is nearby. This means the dem is not the same. I will see what happens if I use the same version of gdal.

I've been trying to make the two environments more similar to achieve more consistent behaviour. This is challenging as python-pdal is not pinned to pdal leading to[ install errors when using pdal=2.2](https://github.com/PDAL/python/issues/84).

I tied pinning all the top level dependencies but got the [install errors when using pdal=2.2](https://github.com/PDAL/python/issues/84).

Then moved to unpin everything and see if we got get tests 1&2 executing even if test 2 had differences in the resulting image.

Getting closer to consistent behaviour - pinned pdal and python pdal. Most DEM values now identical but there are still some that differ. 
![image](https://user-images.githubusercontent.com/22883860/126570281-8eb577c2-b5ce-4d2c-ac49-05fb70e3fd35.png)

Planned next steps?

kinow commented:
I think options are

1. leave PR open until dependencies are sorted out
2. make the test less strict (probably not OK?)
3. mark the tests failing as skipped/ignored and add a blocker issue to fix them before next release
4. merge PR to have the CI running for each commit and ignore the test failures (dangerous?)

rosepearson commented:
> I think options are
> 
> 1. leave PR open until dependencies are sorted out
> 2. make the test less strict (probably not OK?)
> 3. mark the tests failing as skipped/ignored and add a blocker issue to fix them before next release
> 4. merge PR to have the CI running for each commit and ignore the test failures (dangerous?)
> 
> But no idea which one would be the best option for GeoFabrics.

I'm leaning to option 1 or 3 as one of the test's is passing there is still some value in 3. I intend to spend another 1- 2 hrs trying to sorting out dependencies then I will switch to ignoring the failing tests.

kinow commented on tests/test_processor_remote_tiles/test_processor_remote_tiles.py:
```diff
@@ -137,10 +137,12 @@ def test_result_dem(self):
+        print(diff_array[diff_array != 0])
         # compare the generated and benchmark DEMs
-        numpy.testing.assert_array_almost_equal(test_dem.data, benchmark_dem.data,
+        '''numpy.testing.assert_array_almost_equal(test_dem.data, benchmark_dem.data,
```
Is it for disabling the test? If so, it's also possible to use `pytest` decorators for that: https://docs.pytest.org/en/latest/how-to/skipping.html#skipping-test-functions

When the tests are run, the skipped ones will be marked as skipped, instead of passed as it'd be here I think.

rosepearson replied:
Just for now while trying to output the artifacts so I can compare the produced DEMs and debug what is causing the differences - i.e. looks like it might be the tile ordering atm - I will remove before submitting. 

Thanks for the details on how to ignore!

rosepearson replied:
Ok - I've removed it and create a less strict linux test and a stricter windows test.

LukeParky commented on README.md:
```diff
@@ -1,14 +1,20 @@
 The GeoFabrics and associated sub-packages include routines and classes for combining point (i.e. LiDAR), vector (i.e. catchment of interest, infrastructure) and raster (i.e. reference DEM) to generate a hydrologically conditioned raster. 
 
-GeoFabrics also contains support for downloading all LiDAR tiles within a spectified polygon (catchment region) from OpenTopography. This functionality is contained in the lidar_fetch module, and an example of its usage can be found in tests/test_lidar_fetch.
+GeoFabrics also contains support for downloading all LiDAR tiles within a spectified polygon (catchment region) from OpenTopography. This functionality is contained in the lidar_fetch module, and can be used independently of the other geofabrics modules. An example of its usage can be found in tests/test_lidar_fetch.
```
This is really good, do you think this could even be its own package one day? That would have pros of being more modular and would allow people to use it with fewer dependencies. Although it might just make things more complex for you for now. Something to think about post-release maybe

LukeParky commented on README.md:
````diff
@@ -1,14 +1,20 @@
 
 ```python
-conda env create -f environment.yml
+conda env create -f environment_[windows|linux].yml
````
I like the use of this but it might pay to add a comment or use `$BASH_VARIABLES`, whichever you think is the best way to prevent users from thinking the command is literally `conda env create -f environment_[windows|linux].yml` instead of `conda env create -f environment_windows.yml`

rosepearson replied:
Good suggestion I'll have a quick go at clarifying it!

rosepearson replied:
ok I've had a stab at making it clearer in the text above. Let me know what you think.

LukeParky replied:
Looks great, just one quick suggestion then ready to go

LukeParky commented on README.md:
````diff
@@ -1,14 +1,20 @@
 ## import into a conda environment
 You can use this package by using pip to install the package and dependencies using the following commands (say in a conda terminal to add it to that environment), where the environment.yml is from the root of this repository.
 
 ```python
````
```suggestion
```bash
```

LukeParky commented on README.md:
```diff
@@ -31,19 +37,27 @@ In the conda environment defined in the root\environment.yml, run the following:
+## Tests
+Tests exist for stand alone functionality (i.e. fetch lidar), and complete processing chain (i.e. creating a DEM from LiDAR files within a shapefile). A 'benchmark_dem.nc' is uploaded for each test when a DEM is generated. This is stored using git LTS as these file are not human readable. 
+
+### Automated testing
+[Github Actions](https://docs.github.com/en/actions) are used to run tests after each push to remote (i.e. github). [Miniconda](https://github.com/marketplace/actions/setup-miniconda) from the GitHub Actions marketplace is used to install the package dependencies. Linting with [Flake8](https://github.com/py-actions/flake8) and testing with [PyTest](https://docs.pytest.org/en/6.2.x/contents.html)  ...
+
+Check the actions tab after you push to check if your tests run successfully.
```
I like this documentation. Depending on the end goal at some point you may have to look at something bigger than READMEs, perhaps a wiki or some site dedicated to documentation. I don't think this is needed yet though

rosepearson replied:
Great suggestion. Yes - it's definitely getting a bit long! Probably a good documentation issue.

LukeParky commented on README.md:
```diff
@@ -53,16 +67,26 @@ See image below:
+### Running a test
 
-Go to 'Run>Configuration per file...' and check the **The following directory** under **Working directory settings** and specify the root of the repository with no quotes.
+Go to 'Run>Configuration per file...' and check the **The following directory** under **Working directory settings** and specify the root of the repository with no quotes. Doe this for each Python test file (usually one per test folder)
```
These instructions are IDE specific so it would be good to remind the reader which IDE this is for.
Or you could write it IDE-agnostically. Whichever one you prefer

Spelling here:
```suggestion
Go to 'Run>Configuration per file...' and check the **The following directory** under **Working directory settings** and specify the root of the repository with no quotes. Do this for each Python test file (usually one per test folder)
```

LukeParky commented on README.md:
```diff
@@ -53,16 +67,26 @@ See image below:
+Maintainer: Rose Pearson @rosepearson rose.pearson@niwa.co.nz
+
+## License
+[GNU](https://github.com/rosepearson/GeoFabrics/LICENSE)
```
```suggestion
[GNU GPL](https://github.com/rosepearson/GeoFabrics/LICENSE)
```

LukeParky commented on src/geofabrics/processor.py:
```diff
@@ -78,8 +78,8 @@ def get_lidar_file_list(self, verbose) -> list:
             self.lidar_fetcher.run()
-            lidar_file_paths = list(pathlib.Path(self.lidar_fetcher.cache_path /
-                                    self.lidar_fetcher.dataset_prefixes[lidar_dataset_index]).glob('*.laz'))
+            lidar_file_paths = sorted(list(pathlib.Path(self.lidar_fetcher.cache_path /
+                                      self.lidar_fetcher.dataset_prefixes[lidar_dataset_index]).glob('*.laz')))
```
Good idea to sort these but it could be even further simplified since `sorted` will create a `list` out of any `Sequence`

```suggestion
            lidar_file_paths = sorted(pathlib.Path(self.lidar_fetcher.cache_path /
                                      self.lidar_fetcher.dataset_prefixes[lidar_dataset_index]).glob('*.laz'))
```

LukeParky replied:
If you feel this reduces readability then keep it as is

LukeParky replied:
`.glob()` will produce a list so you could also do this:

```suggestion
            lidar_file_paths = pathlib.Path(self.lidar_fetcher.cache_path /
                                      self.lidar_fetcher.dataset_prefixes[lidar_dataset_index]).glob('*.laz')
           lidar_file_paths.sort()                             
```

LukeParky commented on tests/test_processor_remote_tiles/test_processor_remote_tiles.py:
```diff
@@ -124,7 +126,8 @@ def test_correct_file_size(self):
                         f" and the expected sizes of {self.SIZES}")
 
-    def test_result_dem(self):
+    @pytest.mark.skipif(sys.platform != 'win32', reason="Windows test - this is strict")
+    def test_result_dem_windows(self):
```
Good to see pytest has this functionality, good use of annotations

LukeParky commented on src/geofabrics/__init__.py:
```diff
@@ -4,4 +4,4 @@
 """
-__version__ = "0.3.0"
\ No newline at end of file
+__version__ = "0.4.0"
```
Is there a check you can add to the GitHub Actions to bump versions or perform the check. I would never remember to change this

rosepearson replied:
I think I will leave it for now - but your suggestion of using GH actions is great. We could tie it to the associated PR/branch milestone. I'll look into this before we go to release

LukeParky reviewed (changes requested):
Nice one @rosepearson, I can see you're putting a lot of care into the maintainability of this project. I added a couple suggestions so feel free to start a discussion on any of them

rosepearson reviewed (commented):
Thanks for all the commits and comments Luke! Some nice simplification of code in there (i.e. use sort directly).

I think I will leave the version bump manual for now - but I'm sure your right that GH actions can be used to bump it automatically tied to the issue/PR milestone I will leave that for a later time :)

Your changes to the readme are great. I'll add either a comment or $BASHVAR to the readme in regards to the environment_xx.yml.

It's a good point about the use of a wiki. The spyder specific documentation for instance would sit nicely as it's own spyder specific page. It's currently nested under a spyder heading.

LukeParky commented:
> It's a good point about the use of a wiki. The spyder specific documentation for instance would sit nicely as it's own spyder specific page. It's currently nested under a spyder heading.

Ahh of course, I remember the heading but I couldn't scroll high enough in the GitHub diff

LukeParky commented on README.md:
```diff
@@ -10,10 +10,10 @@ The GeoFabrics and associated sub-packages include routines and classes for comb
-## import into a conda environment
-You can use this package by using pip to install the package and dependencies using the following commands (say in a conda terminal to add it to that environment), where the environment.yml is from the root of this repository.
+## Import into a conda environment
+You can use this package by using pip to install the package and dependencies using the following commands in a conda terminal to add it to that environment, where you must either specify environment_windows.yml or environment_linux.yml depending on your operating system. Each file is located in the root repository folder. Sorry there is no macOS support at this stage.
```
```suggestion
You can use this package by using pip to install the package and dependencies using the following commands in a conda terminal to add it to that environment, where you must either specify `environment_windows.yml` or `environment_linux.yml` depending on your operating system. Each file is located in the root repository folder. Sorry there is no macOS support at this stage.
```

rosepearson replied:
Great suggestion!

LukeParky reviewed (changes requested):
Just one formatting suggestion in the README and then it's ready to go. Great PR

LukeParky reviewed (approved):
Good catch
